crypto/types: NewCompactBitArray arithmetic doesn't check that the arguments to make can be negative and be crashed

## Summary of Bug
Auditing this code, you'll see this seemingly boring piece code of code https://github.com/cosmos/cosmos-sdk/blob/bffcae54a1afee8e93468fbf70f7235894b17465/crypto/types/compact_bit_array.go#L19-L27

and there is a bounds check already for `bits <= 0` and then nakedly do the addition https://github.com/cosmos/cosmos-sdk/blob/bffcae54a1afee8e93468fbf70f7235894b17465/crypto/types/compact_bit_array.go#L25
(bits + 7) / 8 is NOT guaranteed to ALWAYS be positive
and alas that can cause a negative to be passed into make simply by passing any value in the range >= maxInt-6 to cause an overflow and here is a test that'll trip
```go
func TestNewCompactBitArrayCrash(t *testing.T) {
        _ = NewCompactBitArray(int(^uint(0) >> 1))
} 
```
which crashes with
```shell
--- FAIL: TestNewCompactBitArrayCrash (0.00s)
panic: runtime error: makeslice: len out of range [recovered]
	panic: runtime error: makeslice: len out of range

goroutine 7 [running]:
testing.tRunner.func1.1(0x1383800, 0x1437510)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc000001e00)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1075 +0x41a
panic(0x1383800, 0x1437510)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/runtime/panic.go:969 +0x1b9
github.com/cosmos/cosmos-sdk/crypto/types.NewCompactBitArray(...)
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/crypto/types/compact_bit_array.go:25
github.com/cosmos/cosmos-sdk/crypto/types.TestNewCompactBitArrayCrash(0xc000001e00)
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/crypto/types/compact_bit_array_test.go:244 +0x45
testing.tRunner(0xc000001e00, 0x13efa68)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1168 +0x2b3
exit status 2
FAIL	github.com/cosmos/cosmos-sdk/crypto/types	0.352s
```
because that value on a 64 bit machine is -9223372036854775802 :-(

## Recommendation
I recommend that we do this 2 fold:
a) check that the addition of (bits + 7) doesn't overflow and is logical
b) add a reasonable limit to bits -- this would have to be a choice left to the discretion of the crypto/types maintainers

from a security perspective let's start with a)
<!-- Concisely describe the issue -->

## Version
Latest at bffcae54a1afee8e93468fbf70f7235894b17465
<!-- git commit hash or release version -->

## Steps to Reproduce

<!-- What commands in order should someone run to reproduce your problem? -->

____

## For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

--- a/crypto/types/compact_bit_array.go
+++ b/crypto/types/compact_bit_array.go
@@ -5,6 +5,7 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
+	"math"
 	"regexp"
 	"strings"
 )
@@ -15,14 +16,24 @@ import (
 // This is not thread safe, and is not intended for concurrent usage.
 
 // NewCompactBitArray returns a new compact bit array.
-// It returns nil if the number of bits is zero.
+// It returns nil if the number of bits is zero, or if there is any overflow
+// in the arithmetic to encounter for the number of its elements: (bits+7)/8,
+// or if the number of elements will be an unreasonably large number like
+// > maxint32 aka >2**31.
 func NewCompactBitArray(bits int) *CompactBitArray {
 	if bits <= 0 {
 		return nil
 	}
+	nElems := (bits + 7) / 8
+	if nElems <= 0 || nElems > math.MaxInt32 {
+		// We encountered an overflow here, and shouldn't pass negatives
+		// to make, nor should we allow unreasonable limits > maxint32.
+		// See https://github.com/cosmos/cosmos-sdk/issues/9162
+		return nil
+	}
 	return &CompactBitArray{
 		ExtraBitsStored: uint32(bits % 8),
-		Elems:           make([]byte, (bits+7)/8),
+		Elems:           make([]byte, nElems),
 	}
 }
 

--- a/crypto/types/compact_bit_array_test.go
+++ b/crypto/types/compact_bit_array_test.go
@@ -2,6 +2,8 @@ package types
 
 import (
 	"encoding/json"
+	"fmt"
+	"math"
 	"math/rand"
 	"testing"
 
@@ -238,4 +240,36 @@ func BenchmarkNumTrueBitsBefore(b *testing.B) {
 			ba.NumTrueBitsBefore(90)
 		}
 	})
+}
+
+func TestNewCompactBitArrayCrashWithLimits(t *testing.T) {
+	if testing.Short() {
+		t.Skip("This test can be expensive in memory")
+	}
+	tests := []struct {
+		in       int
+		mustPass bool
+	}{
+		{int(^uint(0) >> 30), false},
+		{int(^uint(0) >> 1), false},
+		{int(^uint(0) >> 2), false},
+		{int(math.MaxInt32), true},
+		{int(math.MaxInt32) + 1, true},
+		{int(math.MaxInt32) + 2, true},
+		{int(math.MaxInt32) - 7, true},
+		{int(math.MaxInt32) + 24, true},
+		{int(math.MaxInt32) * 9, false}, // results in >=maxint after (bits+7)/8
+		{1, true},
+		{0, false},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(fmt.Sprintf("%d", tt.in), func(t *testing.T) {
+			got := NewCompactBitArray(tt.in)
+			if g := got != nil; g != tt.mustPass {
+				t.Fatalf("got!=nil=%t, want=%t", g, tt.mustPass)
+			}
+		})
+	}
 }

